nnet_utils/nnet_layer.h

```diff
@@ -20,67 +20,85 @@
+    typename CONFIG_T::weight_t  weights[CONFIG_T::n_in][CONFIG_T::n_out],
+    typename CONFIG_T::bias_t    biases[CONFIG_T::n_out])
 {
+    data_T cache;
+    typename CONFIG_T::acc_t mult[CONFIG_T::n_in][CONFIG_T::n_out];
+    typename CONFIG_T::acc_t acc[CONFIG_T::n_out];
 
-    data_T data_cache;
-    acc_T acc[CONFIG_T::n_out];
-
-    // is there a way to cyclically unroll multiple dimensions?
-    #pragma HLS ARRAY_PARTITION variable=weights complete
-    #pragma HLS ARRAY_PARTITION variable=acc complete
-    #pragma HLS ARRAY_PARTITION variable=biases complete
+    // Use a function_instantiate in case it helps to explicitly optimize unchanging weights/biases
+    #pragma HLS function_instantiate variable=weights,biases
 
-    // Optional... Cuts down on a few of the BRAMs
-    // #if CONFIG_T::n_out > 16
-    //     #pragma HLS RESOURCE variable=acc core=RAM_2P_LUTRAM
-    // #endif
+    if (CONFIG_T::io_type == io_parallel){
+        // For parallel inputs:
+        //   - completely partition arrays -- target fabric
+        //   - if we have an unroll factor, limit number of multipliers
+        #pragma HLS PIPELINE
+        #pragma HLS ARRAY_PARTITION variable=weights complete
+        #pragma HLS ARRAY_PARTITION variable=biases complete
+        #pragma HLS ARRAY_PARTITION variable=mult complete
+        #pragma HLS ARRAY_PARTITION variable=acc complete
+        if (CONFIG_T::reuse_factor > 1) {
+            int multiplier_limit  = ceil(CONFIG_T::n_in*CONFIG_T::n_out / CONFIG_T::reuse_factor);
+            #pragma HLS ALLOCATION instances=mul limit=multiplier_limit operation
+        }
+    } else if (CONFIG_T::io_type == io_serial){
+        // TODO: Fill out the directives for serial input
+    }
 
-    int unroll_factor_in  = CONFIG_T::n_in / CONFIG_T::roll_factor_in;
-    int unroll_factor_out = CONFIG_T::n_out / CONFIG_T::roll_factor_out;
-    //int unroll_factor_in  = CONFIG_T::n_in;
-    //int unroll_factor_out = CONFIG_T::n_out;
+    // Do the matrix-multiply
+    Product1: for(int ii = 0; ii < CONFIG_T::n_in; ii++) {
+        cache = data[ii];
+        Product2: for(int jj = 0; jj < CONFIG_T::n_out; jj++) {
+            mult[ii][jj] = cache * weights[ii][jj];
+        }
+    }
 
-    Reset: for(int iacc = 0; iacc < CONFIG_T::n_out; iacc++) {
-      #pragma HLS UNROLL factor=unroll_factor_out
-        acc[iacc] = 0;
+    // Initialize accumulator with input biases
+    ResetAccum: for(int iacc = 0; iacc < CONFIG_T::n_out; iacc++) {
+        acc[iacc] = (typename CONFIG_T::acc_t) biases[iacc];
     }
 
-    NewInput: for(int ii = 0; ii < CONFIG_T::n_in; ii++) {
-        #pragma HLS UNROLL factor=unroll_factor_in
-        data_cache = data[ii];
-        Product: for(int jj = 0; jj < CONFIG_T::n_out; jj++) {
-        #pragma HLS UNROLL factor=unroll_factor_out
-            acc[jj] += data_cache * weights[ii][jj];
+    // Accumulate multiplication result
+    Accum1: for(int ii = 0; ii < CONFIG_T::n_in; ii++) {
```

Comment:
For this 2D loop do we have to do anything special for the resource usage?  The accumulator is not a big worry from the resource-usage side (what is it using, just LUTs?)?

Reply:
Yes, accumulator is just using logic-- it will get big for lots of input features!

I believe for this case we do not need to do anything special. I seem to remember trying a dedicated PIPELINE II=4 to match the resource reuse factor, but the logic consumed was unchanged (might be good to verify). If true, that would suggest the compiler is recognizing the higher II and already capturing that without an explicit directive. 

We may need to add extra pragmas for the serial IO situation